src/v/cluster/plugin_frontend.cc

```diff
@@ -0,0 +1,378 @@
+          if (input_config.is_read_replica()) {
+              return errc::transform_invalid_create;
+          }
+          if (cmd.value.output_topics.empty()) {
+              return errc::transform_invalid_create;
+          }
+          constexpr static size_t max_output_topics = 1;
+          if (cmd.value.output_topics.size() > max_output_topics) {
+              return errc::transform_invalid_create;
+          }
+          absl::flat_hash_set<model::topic_namespace> uniq(
+            cmd.value.output_topics.begin(), cmd.value.output_topics.end());
+          if (uniq.size() != cmd.value.output_topics.size()) {
+              return errc::transform_invalid_create;
+          }
+          for (const auto& out_name : cmd.value.output_topics) {
+              if (out_name.ns != model::kafka_namespace) {
+                  return errc::transform_invalid_create;
+              }
+              auto output_topic = _topics->get_topic_metadata(out_name);
+              if (!output_topic) {
+                  return errc::topic_not_exists;
+              }
+              const auto& output_config = output_topic->get_configuration();
+              if (output_config.is_internal()) {
+                  return errc::transform_invalid_create;
+              }
+              if (output_config.is_read_replica()) {
+                  return errc::transform_invalid_create;
+              }
+              if (
+                output_config.partition_count < input_config.partition_count) {
+                  // copartitioning is required
+                  return errc::transform_invalid_create;
+              }
+              if (would_cause_cycle(cmd.value.input_topic, out_name)) {
+                  return errc::transform_invalid_create;
+              }
+          }
+
+          return errc::success;
+      },
+      [this](const transform_remove_cmd& cmd) {
+          auto transform = _table->find_by_name(cmd.key);
+          if (!transform) {
+              return errc::transform_does_not_exist;
+          }
+          return errc::success;
+      });
+}
+
+bool plugin_frontend::validator::would_cause_cycle(
+  model::topic_namespace_view input, model::topic_namespace output) {
+    model::topic_namespace_eq eq;
+    // Does output ever lead to input? let's find out via breadth first search.
+    ss::circular_buffer<model::topic_namespace> queue;
+    queue.push_back(std::move(output));
+    // TODO: should we bound this loop and if you exceed it then we just say
+    // you've chained too many functions together?
+    while (!queue.empty()) {
+        model::topic_namespace current = std::move(queue.front());
+        queue.pop_front();
+        if (eq(input, current)) {
+            return true;
+        }
+        auto metas = _table->find_by_input_topic(current);
+        for (auto& [id, meta] : metas) {
+            for (auto& output_topic : meta.output_topics) {
+                queue.push_back(std::move(output_topic));
```

Comment:
move of lvalue?

Reply:
removed it, I was just stealing the strings out of `metas` (which is a local) to prevent some copies